scipy/optimize/_linprog_util.py

```diff
@@ -128,16 +180,20 @@ def _clean_inputs(c, A_ub=None, b_ub=None, A_eq=None, b_eq=None, bounds=None,
         Starting values of the independent variables, which will be refined by
         the optimization algorithm.
     """
+    if c is None:
+        raise TypeError
 
     try:
-        if c is None:
-            raise TypeError
-        try:
-            c = np.asarray(c, dtype=float).copy().squeeze()
-        except BaseException:  # typically a ValueError and shouldn't be, IMO
-            raise TypeError
+        c = np.array(c, dtype=np.float, copy=True).squeeze()
+    except ValueError:
+        raise TypeError(
+            "Invalid input for linprog: c must be a 1D array of numerical "
+            "coefficients")
+    else:
```

Comment:
good call. 
`else` after `except` wasn't in my repertoire when I wrote this, but it looks like it accomplishes what I was after.